src/data/fun_like.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2021 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+
+import tactic.lint
+import tactic.norm_cast
+
+/-!
+# Typeclass for a type `F` with an injective map to `A → B`
+
+This typeclass is primarily for use by homomorphisms like `monoid_hom` and `linear_map`.
+
+## Basic usage of `fun_like`
+
+A typical type of morphisms should be declared as:
+```
+structure my_hom (A B : Type*) [my_class A] [my_class B] :=
+(to_fun : A → B)
+(map_op' : ∀ {x y : A}, to_fun (my_class.op x y) = my_class.op (to_fun x) (to_fun y))
+
+namespace my_hom
+
+variables (A B : Type*) [my_class A] [my_class B]
+
+-- This instance is optional if you follow the "Hom class" design below:
+instance : fun_like (my_hom A B) A B :=
+{ coe := my_hom.to_fun, coe_injective' := λ f g h, by cases f; cases g; congr' }
+
+/-- Helper instance for when there's too many metavariables to apply `to_fun.to_coe_fn` directly. -/
+instance : has_coe_to_fun (M →ₗ[R] M₂) := to_fun.to_coe_fn
```

Comment:
Are you intentionally talking about linear maps in this example? Or did you mean `my_hom`?

Reply:
Oops, copy-paste error.
```suggestion
instance : has_coe_to_fun (my_hom A B) := to_fun.to_coe_fn
```